sklearn/cluster/_kmeans.py

```diff
@@ -897,10 +900,19 @@ def _check_params_vs_input(self, X, default_n_init=None):
             )
             self._n_init = default_n_init
         if self._n_init == "auto":
-            if self.init == "k-means++":
+            if isinstance(self.init, str) and self.init == "k-means++":
                 self._n_init = 1
-            else:
+            elif isinstance(self.init, str) and self.init == "random":
+                self._n_init = default_n_init
+            elif callable(self.init):
                 self._n_init = default_n_init
+            elif _is_arraylike_not_scalar(self.init):
+                self._n_init = 1
+            else:
+                raise ValueError(
+                    'Expect init to be one of ["k-means++", "random", callable or'
+                    "array-like of shape(n_clusters, n_features)]"
+                )
```

Comment:
The check that `init` has a valid type and value is already taken care of by a generic validation tool.
```suggestion
            else:  # array-like
                self._n_init = 1
```